scipy/sparse/linalg/_expm_multiply.py

```diff
@@ -41,6 +44,53 @@ def _trace(A):
         return np.trace(A)
 
 
+def traceest(A, m3, seed=None):
+    """Estimate `np.trace(A)` using `3*m3` matrix-vector products.
+
+    The result is not deterministic.
+
+    Parameters
+    ----------
+    A : LinearOperator
+        Linear operator whose trace will be estimated. Has to be square.
+    m3 : int
+        Number of matrix-vector products divided by 3 used to estimate the
+        trace.
+    seed : optional
+        Seed for `numpy.random.default_rng`.
+        Can be provided to obtain deterministic results.
+
+    Returns
+    -------
+    trace : LinearOperator.dtype
+        Estimate of the trace
+
+    Notes
+    -----
+    This is the Hutch++ algorithm given in [1]_.
+
+    References
+    ----------
+    .. [1] Meyer, Raphael A., Cameron Musco, Christopher Musco, and David P.
+       Woodruff. "Hutch++: Optimal Stochastic Trace Estimation." In Symposium
+       on Simplicity in Algorithms (SOSA), pp. 142-155. Society for Industrial
+       and Applied Mathematics, 2021
+       https://doi.org/10.1137/1.9781611976496.16
+
+    """
+    rng = np.random.default_rng(seed)
+    if len(A.shape) != 2 or A.shape[-1] != A.shape[-2]:
+        raise ValueError("Expected A to be like a square matrix.")
+    n = A.shape[-1]
+    S = rng.choice([-1.0, +1.0], [n, m3])
+    Q, _ = qr(A.matmat(S), overwrite_a=True, mode='economic')
```

Comment:
What's the typical size of a LinearOperator to use this function? A dense matrix with a sparse size will definitely kill this line with a MemoryError.

Reply:
Of course, we should allow as big `LinearOperator`s `A` as feasible. Where do you see the problem (sorry, I am not experienced enough yet)?

For ``A.shape=(N, N)``, the matrix ``A.matmat(S)`` is of size ``(N, m3)``. For any algorithm, we have to be able to save at the very least a vector of size ``(N, )`` in memory. Here, this is `m3` times that size. So the issue should be with QR, right?
``Q.shape=(N, m3)`` and ``R.shape=(m3, m3)`` should still be feasible. The Housholder QR needs ``2 * m3**2 * ( N - m3/3)`` flops (Golub & Loan) which still seems feasible (prefactor `3` for Givens QR).
However, I don't know the memory requirements.

To my knowledge, the main restriction for `m3` is that the QR costs O(`N * m3**2`).

Reply:
If the array size is more than 5k we are in the territory of memory errors here. Because A might be sparse with very small nnz but Q will be full from this computation.

Reply:
@ilayn true, but no way out of this, since `A.matmat(S)` will already be dense because `S` is dense. Estimating the trace is not so cheap in this algorithm.